indexing-service/src/main/java/io/druid/indexing/common/task/ParallelIndexSubTask.java

```diff
@@ -0,0 +1,451 @@
+  }
+
+  private boolean checkLockAcquired(TaskActionClient actionClient, SortedSet<Interval> intervals)
+  {
+    try {
+      tryAcquireExclusiveSurrogateLocks(actionClient, intervals);
+      return true;
+    }
+    catch (Exception e) {
+      log.error(e, "Failed to acquire locks for intervals[%s]", intervals);
+      return false;
+    }
+  }
+
+  @JsonProperty
+  public int getNumAttempts()
+  {
+    return numAttempts;
+  }
+
+  @JsonProperty("spec")
+  public ParallelIndexIngestionSpec getIngestionSchema()
+  {
+    return ingestionSchema;
+  }
+
+  @JsonProperty
+  public String getSupervisorTaskId()
+  {
+    return supervisorTaskId;
+  }
+
+  @Override
+  public TaskStatus run(final TaskToolbox toolbox) throws Exception
+  {
+    final boolean explicitIntervals = ingestionSchema.getDataSchema()
+                                                     .getGranularitySpec()
+                                                     .bucketIntervals()
+                                                     .isPresent();
+
+    final FirehoseFactory firehoseFactory = ingestionSchema.getIOConfig().getFirehoseFactory();
+
+    if (firehoseFactory instanceof IngestSegmentFirehoseFactory) {
+      // pass toolbox to Firehose
+      ((IngestSegmentFirehoseFactory) firehoseFactory).setTaskToolbox(toolbox);
+    }
+
+    final File firehoseTempDir = toolbox.getFirehoseTemporaryDir();
+    // Firehose temporary directory is automatically removed when this IndexTask completes.
+    FileUtils.forceMkdir(firehoseTempDir);
+
+    final DataSchema dataSchema = ingestionSchema.getDataSchema();
+    final Map<Interval, String> versions;
+
+    if (explicitIntervals) {
+      versions = toolbox.getTaskActionClient().submit(new SurrogateAction<>(supervisorTaskId, new LockListAction()))
+                        .stream()
+                        .collect(Collectors.toMap(TaskLock::getInterval, TaskLock::getVersion));
+    } else {
+      versions = null;
+    }
+
+    final List<DataSegment> pushedSegments = generateAndPushSegments(
+        toolbox,
+        dataSchema,
+        versions,
+        firehoseFactory,
+        firehoseTempDir
+    );
+    final ParallelIndexTaskClient taskClient = taskClientFactory.build(
+        new ClientBasedTaskInfoProvider(indexingServiceClient),
+        getId(),
+        1, // always use a single http thread
+        ingestionSchema.getTuningConfig().getChatHandlerTimeout(),
+        ingestionSchema.getTuningConfig().getChatHandlerNumRetries()
+    );
+    taskClient.report(supervisorTaskId, pushedSegments);
+
+    return TaskStatus.success(getId());
+  }
+
+  private Map<Interval, TaskLock> tryAcquireExclusiveSurrogateLocks(
+      TaskActionClient client,
+      SortedSet<Interval> intervals
+  )
+      throws IOException
+  {
+    final Map<Interval, TaskLock> lockMap = new HashMap<>();
+    for (Interval interval : Tasks.computeCompactIntervals(intervals)) {
```

Comment:
This can be done in a fluent manner using the Collectors.toMap

Reply:
Actually, the return value is never consumed. Do you need to build the map?

Reply:
Good point. Removed.